rpc/src/rpc.rs

```diff
@@ -1857,29 +1857,123 @@ impl JsonRpcRequestProcessor {
+                amount: Reverse<u64>,
+
+                inner: RpcTokenAccountBalance,
+            }
+            /// We only need partially ordered entries (total by u64)
+            impl PartialOrd for RpcTokenAccountBalanceWrapper {
+                fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+                    // Reverse wrapper reverses this cmp to make maxheap a minheap
+                    Some(self.amount.cmp(&other.amount))
+                }
+            }
+            impl Eq for RpcTokenAccountBalanceWrapper {}
+            impl Ord for RpcTokenAccountBalanceWrapper {
+                fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+                    self.amount.cmp(&other.amount)
+                }
+            }
+
+            // Pre allocate minheap
+            let mut token_accounts_min_heap = BinaryHeap::with_capacity(NUM_LARGEST_ACCOUNTS);
+
+            for (address, account) in token_accounts.into_iter() {
+                let amount_u64 = StateWithExtensions::<TokenAccount>::unpack(account.data())
                     .map(|account| account.base.amount)
                     .unwrap_or(0);
-                let amount = token_amount_to_ui_amount(amount, decimals);
-                RpcTokenAccountBalance {
-                    address: address.to_string(),
-                    amount,
+
+                // Create a wrapped type that is used for the min heap
+                let mut wrapped_token_account_balance = RpcTokenAccountBalanceWrapper {
+                    inner: RpcTokenAccountBalance {
+                        // Only do base58 if we are adding to heap and returning
+                        address: String::new(),
+                        // Only allocate strings if we are adding to heap and returning
+                        amount: UiTokenAmount {
+                            ui_amount: Default::default(),
+                            decimals,
+                            amount: Default::default(),
+                            ui_amount_string: Default::default(),
+                        },
+                    },
+                    amount: Reverse(amount_u64),
+                };
+
+                // Auxiliary booleans for evict_one & add_to_heap conditions
+                let is_not_full = token_accounts_min_heap.len() < NUM_LARGEST_ACCOUNTS;
+                // Reverse requires < to check for larger, e.g. Reverse(5) < Reverse(3).
+                // Note that Some(...) < None evaluates to false, but add_to_heap will be true
+                // because None is only returned when the heap is empty (is_not_full = true)
+                let smallest_in_heap = token_accounts_min_heap.peek();
+                let is_larger = Some(&wrapped_token_account_balance) < smallest_in_heap;
+
+                // We evict smallest entry if the current entry is larger AND the heap is full
+                let evict_one = is_larger & !is_not_full;
```

Comment:
any reason this is a bitwise and?

feels like it should be `is_larger && !is_not_full`

Reply:
I usually default to bitwise and for better branch predicting.

Reply:
looks like it works fine, ^ just a nit

Reply:
I'll remove all the bitwise ops then. There's a bitwise or in there as well